scripts/update-message-registry.py

```diff
@@ -0,0 +1,205 @@
+            type=type_,
+            description=description,
+            added=added,
+            removed=removed,
+            details=details,
+            examples=examples,
+        )
+
+    def message_code(self) -> str:
+        """Format the message code for display"""
+        return f"XKB-{self.code:0>3}"
+
+    def message_index(self: Entry) -> str:
+        """Returns the C enumeration member denoting the message index"""
+        id = self.id.replace("-", "_").upper()
+        return f"XKB_{self.type.upper()}_{id}"
+
+    def message_name(self: Entry):
+        """Format the message string identifier for display"""
+        return self.id.replace("-", " ").capitalize()
+
+
+def prepend_todo(text: str) -> str:
+    if text.startswith("TODO"):
+        return f"""<span class="todo">{text[:5]}</span>{text[5:]}"""
+    else:
+        return text
+
+
+def generate(
+    registry: Sequence[Entry],
+    env: jinja2.Environment,
+    root: Path,
+    file: Path,
+    skip_removed: bool = False,
+):
+    """Generate a file from its Jinja2 template and the message registry"""
+    template_path = file.with_suffix(f"{file.suffix}.jinja")
+    template = env.get_template(str(template_path))
+    path = root / file
+    with path.open("wt", encoding="utf-8") as fd:
+        entries = (
+            tuple(filter(lambda e: e.removed is None, registry))
+            if skip_removed
+            else registry
+        )
+        fd.writelines(template.generate(entries=entries))
+
+
+# Root of the project
+ROOT = Path(__file__).parent.parent
+
+# Parse commands
+parser = argparse.ArgumentParser(description="Generate files from the message registry")
+parser.add_argument(
+    "--root",
+    type=Path,
+    default=ROOT,
+    help="Path to the root of the project (default: %(default)s)",
+)
+
+args = parser.parse_args()
+
+message_registry_path = args.root / "doc" / "message-registry.yaml"
+
+# Load message registry
+with message_registry_path.open("rt", encoding="utf-8") as fd:
+    # NOTE: sort for cleaner output
+    message_registry = sorted(
+        map(Entry.parse, yaml.safe_load(fd)), key=lambda e: e.code
+    )
+
+# Check message codes and identifiers are unique
+codes: set[int] = set()
+identifiers: set[str] = set()
+for n, entry in enumerate(message_registry):
+    if entry.code in codes:
+        raise ValueError("Duplicated code in entry #{n}: {entry.code}")
+    if entry.id in identifiers:
+        raise ValueError("Duplicated identifier in entry #{n}: {entry.id}")
+    codes.add(entry.code)
+    identifiers.add(entry.id)
+
+# Configure Jinja
+template_loader = jinja2.FileSystemLoader(args.root, encoding="utf-8")
+jinja_env = jinja2.Environment(
+    loader=template_loader,
+    keep_trailing_newline=True,
+    trim_blocks=True,
+    lstrip_blocks=True,
+)
+jinja_env.filters["message_index"] = Entry.message_index
+jinja_env.filters["message_code"] = Entry.message_code
+jinja_env.filters["message_name"] = Entry.message_name
```

Comment:
jinja shouldn't need separate filters for these, you can use `{{entry.message_index}}` and pretend it's just an attribute. That requires `@property` in the class but otherwise it'll just work. Or it should, anyway :)

Reply:
I see, fixed. I use to know these things…